clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java

```diff
@@ -33,139 +31,89 @@
+                    OffsetFetchRequestTopic topic = offsetFetchRequestTopicMap.getOrDefault(
+                        topicName, new OffsetFetchRequestTopic().setName(topicName));
+                    topic.partitionIndexes().add(topicPartition.partition());
+                    offsetFetchRequestTopicMap.put(topicName, topic);
+                }
+            }
+
+            List<OffsetFetchRequestTopic> topics = requestAllPartitions ?
+                                                       new ArrayList<>(offsetFetchRequestTopicMap.values()) : ALL_TOPIC_PARTITIONS;
+            this.data = new OffsetFetchRequestData()
+                            .setGroupId(groupId)
+                            .setTopics(topics);
         }
 
         public static Builder allTopicPartitions(String groupId) {
-            return new Builder(groupId, ALL_TOPIC_PARTITIONS);
+            return new Builder(groupId, null);
         }
 
         public boolean isAllTopicPartitions() {
-            return this.partitions == ALL_TOPIC_PARTITIONS;
+            return this.data.topics() == ALL_TOPIC_PARTITIONS;
         }
 
         @Override
         public OffsetFetchRequest build(short version) {
             if (isAllTopicPartitions() && version < 2)
                 throw new UnsupportedVersionException("The broker only supports OffsetFetchRequest " +
                         "v" + version + ", but we need v2 or newer to request all topic partitions.");
-            return new OffsetFetchRequest(groupId, partitions, version);
+            return new OffsetFetchRequest(data, version);
         }
 
         @Override
         public String toString() {
-            StringBuilder bld = new StringBuilder();
-            String partitionsString = partitions == null ? "<ALL>" : Utils.join(partitions, ",");
-            bld.append("(type=OffsetFetchRequest, ").
-                    append("groupId=").append(groupId).
-                    append(", partitions=").append(partitionsString).
-                    append(")");
-            return bld.toString();
+            return data.toString();
+        }
+    }
+
+    public List<TopicPartition> partitions() {
+        if (isAllPartitions()) {
+            return null;
+        }
+        List<TopicPartition> partitions = new ArrayList<>();
+        for (OffsetFetchRequestTopic topic : data.topics()) {
+            for (Integer partitionIndex : topic.partitionIndexes()) {
+                partitions.add(new TopicPartition(topic.name(), partitionIndex));
+            }
         }
+        return partitions;
     }
 
-    private final String groupId;
-    private final List<TopicPartition> partitions;
+    public String groupId() {
+        return data.groupId();
+    }
 
     public static OffsetFetchRequest forAllPartitions(String groupId) {
-        return new OffsetFetchRequest.Builder(groupId, null).build((short) 2);
+        return OffsetFetchRequest.Builder.allTopicPartitions(groupId).build((short) 2);
```

Comment:
The dependence on version 2 is strange. Since this method is just used in test cases, maybe we can remove it and just use `Builder.allTopicPartitions` directly.